src/sys/windows/tty.rs

```diff
@@ -131,76 +134,79 @@ pub fn set_raw_input_mode(enable: bool) -> bool {
+        return false;
+    };
 
-pub fn is_tty<T: AsRawHandle>(stream: &T) -> bool {
-    let stream = stream.as_raw_handle() as HANDLE;
-    return stream != INVALID_HANDLE_VALUE && unsafe { GetFileType(stream) } == FILE_TYPE_CHAR;
+    return true;
 }
 
+#[derive(Debug)]
 struct WindowsConIn {
-    handle: HANDLE,
-    buffered_utf8: Vec<u8>,
+    handle        : HANDLE,
+    buffered_utf8 : Vec<u8>,
     buffered_utf16: Vec<u16>,
 }
 
 impl WindowsConIn {
-    const MAX_BYTES_TO_READ: usize = 8192;
+    const MAX_BYTES_TO_READ      : usize = 8192;
     const MAX_UTF16_CHARS_TO_READ: usize = WindowsConIn::MAX_BYTES_TO_READ / 3;
 
     fn new() -> io::Result<WindowsConIn> {
         // UTF-16 encoded CONIN$ file
-        let conin_file: Vec<u16> = "CONIN$\0".encode_utf16().collect();
-        let con_in_handle: HANDLE = unsafe { CreateFileW(conin_file.as_ptr(), GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ, null_mut(), OPEN_EXISTING, 0, null_mut()) };
+        let conin_file   : Vec<u16> = "CONIN$\0".encode_utf16().collect();
+        let con_in_handle: HANDLE   = unsafe { 
+            CreateFileW(conin_file.as_ptr(), 
+                GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ, null_mut(), OPEN_EXISTING, 0, null_mut()
+            ) 
+        };
 
         if con_in_handle == INVALID_HANDLE_VALUE {
             Err(::std::io::Error::last_os_error()) // TODO: Figure out how to get this error
-        }
-        else {
+        } else {
             Ok(WindowsConIn {
-                handle: con_in_handle,
+                handle        : con_in_handle,
                 buffered_utf16: Vec::with_capacity(WindowsConIn::MAX_UTF16_CHARS_TO_READ),
-                buffered_utf8: Vec::with_capacity(WindowsConIn::MAX_BYTES_TO_READ),
+                buffered_utf8 : Vec::with_capacity(WindowsConIn::MAX_BYTES_TO_READ),
             })
         }
     }
 
     fn buffer_into_utf8(&mut self) -> io::Result<()> {
         let hconin = &self.handle;
 
-        loop {
-            let mut utf_16_chars_read: DWORD = 0;
-            let succeeded: BOOL = unsafe { ReadConsoleW(*hconin, (self.buffered_utf16.as_mut_ptr() as LPWSTR) as LPVOID, WindowsConIn::MAX_UTF16_CHARS_TO_READ as DWORD, &mut utf_16_chars_read, null_mut()) };
-
-            if succeeded == FALSE {
-                Err(::std::io::Error::last_os_error())?;
-            }
-
-            if utf_16_chars_read == 0 {
-                break;
-            }
-            else {
-                unsafe { self.buffered_utf16.set_len(utf_16_chars_read as usize); }
-                let utf8_from_console = String::from_utf16_lossy(&self.buffered_utf16);
-
-                assert!(utf8_from_console.len() + self.buffered_utf8.len() < self.buffered_utf8.capacity());
-                // XXX: Could probably optimize better with WideStringToMultiByte
-                self.buffered_utf8.extend_from_slice(&utf8_from_console.into_bytes()[..]);
-            }
+        self.buffered_utf16.clear();
```

Comment:
We were wrong to think that if there are 10 bytes in the console stream and we only read 5, then the rest of the bytes are discarded. I found out that, ReadConsole will read x number of bytes and then remove the read x number of bytes from the stream. The next iteration will read and return the remaining bytes. So a loop is actually not needed.

Since WindowsConIn is only initialized once by async_stdin(), the contents of the buffer are never cleared. So if we don't clear the buffer, when we run async.exe and press the 'a' key, the output will be 97 (ascii key code). If we then press the 'b' key, the output will be 97,98. So, we need to clear this buffer every time the thread wakes up otherwise it holds on to data that has already been processed.